src/Services/Air/Air.js

```diff
@@ -134,19 +135,28 @@ module.exports = (settings) => {
     },
 
     ticket(options) {
-      return this.getPNR(options).then((booking) => {
-        const ticketParams = Object.assign({}, options, {
-          ReservationLocator: booking.uapi_reservation_locator,
-        });
-        return service.ticket(ticketParams)
-          .then(result => result, (err) => {
-            if (err instanceof AirRuntimeError.TicketingFoidRequired) {
-              return this.getPNR(options)
-                .then(updatedBooking => service.foid(updatedBooking))
-                .then(() => service.ticket(ticketParams));
-            }
-            return Promise.reject(err);
-          });
+      return retry({ retries: 3 }, (again, number) => {
+        if (settings.debug && number > 1) console.log(`ticket ${options.pnr} retry number ${number}`);
+        return (options.ReservationLocator ? Promise.resolve(options) :
+          this.getPNR(options).then((booking) => {
+            const ticketParams = Object.assign({}, options, {
+              ReservationLocator: booking.uapi_reservation_locator,
+            });
+            return ticketParams;
+          })
+        ).then(
+          ticketParams => service.ticket(ticketParams)
+            .then(result => result, (err) => {
+              if (err instanceof AirRuntimeError.TicketingFoidRequired) {
+                return this.getPNR(options)
+                  .then(updatedBooking => service.foid(updatedBooking))
+                  .then(() => this.ticket(ticketParams));
```

Comment:
Shouldn't be `service.ticket` here?
Otherwise it will be retry in retry.

Reply:
No, I specifically changed it to do a global retry.
Global retry (with optional `ReservationLocator` fetch) instead of an internal one from `service` allows not only simple retry scenarios like PNRBusy->Ticketing, but more complex one like FOID+Busy+Ticketing (depth limited by retry count).

Try changing it into `service` to see a unit test fail.

Reply:
This gives you more retries, because it's retry inside retry.
I believe this should be changed to:

```javascript
return this.getPNR(options)
  .then(updatedBooking => service.foid(updatedBooking))
  .then(() => again(err));
```

This will make possible for `AirRuntimeError.TicketingFoidRequired` and `AirRuntimeError.TicketingPNRBusy` go one by one.